src/StatsdClient/StatsdBuilder.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            _factory = factory;
+        }
+
+        public Statsd BuildStats(StatsdConfig config)
+        {
+            var statsdServerName = !string.IsNullOrEmpty(config.StatsdServerName)
+                    ? config.StatsdServerName
+                    : Environment.GetEnvironmentVariable(StatsdConfig.DD_AGENT_HOST_ENV_VAR);
+
+            if (string.IsNullOrEmpty(statsdServerName))
+            {
+                throw new ArgumentNullException(
+                    $"{nameof(config)}.{nameof(config.StatsdServerName)} and"
+                    + $" {StatsdConfig.DD_AGENT_HOST_ENV_VAR} environment variable not set");
+            }
+
+            var metricsSender = CreateMetricsSender(config, statsdServerName);
+            var statsD = new Statsd(metricsSender,
+                                    new RandomGenerator(),
+                                    new StopWatchFactory(),
+                                    "",
+                                    config.ConstantTags);
+            statsD.TruncateIfTooLong = config.StatsdTruncateIfTooLong;
+            return statsD;
+        }
+
+        IStatsdUDP CreateMetricsSender(StatsdConfig config, string statsdServerName)
+        {
+            int bufferCapacity;
+            if (statsdServerName.StartsWith(UnixDomainSocketPrefix))
+            {
+                statsdServerName = statsdServerName.Substring(UnixDomainSocketPrefix.Length);
+                var endPoint = new UnixEndPoint(statsdServerName);
+                _statsSender = _factory.CreateUnixDomainSocketStatsSender(endPoint,
+                                                                          config.Advanced.UDSBufferFullBlockDuration);
+                bufferCapacity = config.StatsdMaxUnixDomainSocketPacketSize;
+            }
+            else
+            {
+                _statsSender = CreateUDPStatsSender(config, statsdServerName);
+                bufferCapacity = config.StatsdMaxUDPPacketSize;
+            }
+
+            _statsBufferize = CreateStatsBufferize(_statsSender,
+                                                   bufferCapacity,
+                                                   config.Advanced);
+            return _statsBufferize;
+        }
+
+        StatsBufferize CreateStatsBufferize(
+            StatsSender statsSender,
+            int bufferCapacity,
+            AdvancedStatsConfig config)
+        {
+            var bufferBuilder = new BufferBuilder(statsSender, bufferCapacity, "\n");
+
+            var statsBufferize = _factory.CreateStatsBufferize(
+                bufferBuilder,
+                config.MaxMetricsInAsyncQueue,
+                config.MaxBlockDuration,
+                config.DurationBeforeSendingNotFullBuffer);
+
+            return statsBufferize;
+        }
+
+        StatsSender CreateUDPStatsSender(StatsdConfig config, string statsdServerName)
+        {
+            var address = StatsdUDP.GetIpv4Address(statsdServerName);
```

Comment:
I have kept the existing behavior https://github.com/DataDog/dogstatsd-csharp-client/blob/master/src/StatsdClient/StatsdUDP.cs#L67-L94.
I am not familiar with DNS resolution. Do you know how can I detect the right ip adresse to use, how to test it and how to avoid breaking change?

Reply:
Let's keep the current implementation. We can discuss about ipv6 support in follow-up discussions. I think it applies to multiple clients.